File: java/v1/src/main/java/com/api/v1/services/rent/RentTerminationServiceImpl.java
package com.api.v1.services.rent;

import com.api.v1.domain.rent.RentRepository;
import com.api.v1.exceptions.rent.TerminatedRentException;
import com.api.v1.utils.rent.RentFinderUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class RentTerminationServiceImpl implements RentTerminationService {

    @Autowired
    private RentRepository repository;

    @Autowired
    private RentFinderUtil rentFinderUtil;

    @Override
    public Mono<Void> terminate(String orderNumber) {
        return rentFinderUtil
                .find(orderNumber)
                .flatMap(rent -> {
                    if (rent.getReturnedAt() != null) {
                        return Mono.error(new TerminatedRentException(orderNumber));
                    }
                    rent.terminate();
                    return repository.save(rent);
                })
                .then();
    }

}

File: v2/src/main/kotlin/com/api/v2/rent/service/RentTerminationServiceImpl.kt
package com.api.v2.rent.service

import com.api.v2.rent.domain.RentRepository
import com.api.v2.rent.exceptions.TerminatedRentException
import com.api.v2.rent.utils.RentFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class RentTerminationServiceImpl: RentTerminationService {

    @Autowired
    lateinit var rentFinderUtil: RentFinderUtil

    @Autowired
    lateinit var rentRepository: RentRepository

    override suspend fun terminate(orderNumber: String) {
        return withContext(Dispatchers.IO) {
            val existingRent = rentFinderUtil.find(orderNumber)
            if (existingRent.returnedAt.isNotEmpty()) {
                throw TerminatedRentException(orderNumber)
            }
            val terminatedRent = existingRent.markAsReturned()
            rentRepository.save(terminatedRent)
        }
    }

}